src/language-js/parse/postprocess/index.js

```diff
@@ -173,6 +173,13 @@ function postprocess(ast, options) {
           });
         }
         break;
+      case "TSUnionType":
+        if (node.types.length === 1 && (node.types[0].type === "TSUnionType" || node.types[0].type === "TSIntersectionType")) {
+          node.types[0].loc = node.loc
+          node.types[0].range = node.range
+          return node.types[0]
+        }
```

Comment:
This is not correct. But before you fix it. Please move tests to a new dir in current one, and add `flow` parser to `jsfmt.spec.js`, and add some comments in code, you'll see what I mean.

Reply:
I've added `loc` and `range` expanding to prevent comments detaching

```tsx
// Input
type A6  = /*1*/ | C;
type D1 = /*1*/ | a & b;

// Output without loc and range inheritting
type A6 /*1*/ = C;
type D1 /*1*/ = a & b;

// Output with inheritting (as expected)
type A6 = /*1*/ C;
type D1 = /*1*/ a & b
```

How can I avoid detaching comments without inheritting `loc` and `range` from parent type?

Reply:
I'm aware of that, but that's how `flow` works, [playground](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAggNm2wF5sB6AKgEYKzsAfbAYQG4AdWTHAESpPOq0G2AIbYAZNgBGLEABoQEDDACW0AM7JQIgE46IAdwAKuhJpQiANgZFpNCqTpFgA1nBgBlEQFs4AGRUoOGQAMyt1OAcnV3cPDGdAgHNkGB0AV0iQOG8pOAATPPy-EShEtJFEuAAxCB1vERhVUuQQETSYCHkQAAsYb0sAdW6VeHV4sDgPMxGVADcRtBawdXsQQIidGCMnRPrQ8MyAK3UADw8kyzgARTSIeH3LCIV4nQ2WkMtDLowdQJgBlR5GDdZAADgADM99BEBk4MC0fnANrNggoAI63eDbJTmVrqAC0QXy+S6OjgGJUZO2FT2SDCj0yEW8KhS6UZFzguEavyk7TgRjgOgCQQeTxA6g5NzuwTpBwUMBEUgBQJBSAATPKnCpLEkmBBvLSsuoAKxdNIRAAqivM9LFswyAEkoIVYB4wL9lLhnR50JdRXAAL4BoA),

We'll fix that later. First make it print consistently with flow parser.

Reply:
If I add 
```tsx
case "TSUnionType":
  if (node.types.length === 1) {
    return node.types[0];
  }
  break;
```
to the `src/language-js/parse/postprocess/index.js`, then `typescript` and `babel-ts` parsers incorrectly attach comments. They attach comments to the `TSTypeAliasDeclaration` in those cases:
https://github.com/prettier/prettier/blob/f602be5a478a5bcdca870b31936dc873ecae1120/tests/format/typescript/union/union-parens.ts#L91
https://github.com/prettier/prettier/blob/f602be5a478a5bcdca870b31936dc873ecae1120/tests/format/typescript/intersection/intersection-parens.ts#L38
So comments are placed before the `=` sign in type declarations

----

But if I'll bring union check back to the 
https://github.com/prettier/prettier/blob/f602be5a478a5bcdca870b31936dc873ecae1120/src/language-js/print/type-annotation.js#L94
all is fine, all tests pass. And it doesn't affect `flow` at all. (Btw `flow` prints all edge cases [correctly](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAggRmwF4AdKbbAH2wAoyKLq7yHHth7WuBrONJbAGcYAJwCWUAOYBuTl2wBfOdgCUnFbNiYcuAEzFOTZdQ4t5vfkNESZyikrNV2dhhYFQArgFsARnBGa8g4UGmToWHgAzAYsTE6mrG5W4lKBiqqa4ToALDFsNE4FJspJwim2LAqqGWHaeACseYXGtC4m2KXWUopt7B18AmU2PY5swP2WQ93BXO2d5SOsao5zA8nDM2MTg12SizVaEbgAbMTNLAD0AFT4VxfnrMWj2+upyjPLIZq1EQBChKRYq1HAAyYHyBLybAANwAhgAbDxwHblNKsTaqdSZOq-fSA-LKMHMeRGZ4UIkuViQqEMOGI5GvCo0+yUzHPT4MDkaEAAGhAEAwMDE0EEyFAsJEIggAHcAAoShCilAI6WwtCivk+ESwsC8GAAZVhXjgABkJHBkAAzBGCOCa7W6uAGjA6mzIURIvlwXxwAAmvr9JthUg8sMkcAAYhARF5YTAhVJkCBYR4YBBeSAABYwLzwgDqmbE8EELrAcH1iqLYmhRbQSbAgg1IAktpEMFl2sksatNrtIAAVoIAB76mzwuAARQ8EHgPfhtr5LpEraTlvhMozGBSMDzYl9MEzyAAHAAGRdS2157UYJNbuCt6EWvkAR2n8A7AqVycEAFooHA-T9DMRDgV8xBAjsw27JBrXnPtbS8MR3RET0QEEMc4FweNxB8VM4FlfwzX-OcFzQjCpxnC0YN7PkYFhHxd33Q8kF0WjtTEeEbAAYQgLxoJAe96gzDxbQAFXopVYNI6EkQASSgANYH1MBxEFXAFP1dBxxIuAFAUIA)) 

----

Maybe I need to perform a check in function `printIntersectionType`?
Because as I understand it if I'll strip redundant unions in `postprocess` without `loc` and `range` inheritance it'll require to change the processing of comments to handle this particular case

Reply:
Since in flow parser we can't know where the "redundant" `|`  is, and it's not breaking syntax, only print comments incorrectly.
I think the goal in this PR should be make the output same as flow parser, even the comments print in unexpected position.
After we get it done, we 'll deal with the comment issue together (both `typescript` and `flow` parser). Can be in a follow-up PR.

Reply:
Yea, sure, sounds great! I will commit fixes soon